components/layout_2020/flow/inline/mod.rs

```diff
@@ -2345,31 +2350,33 @@ impl<'layout_data> ContentSizesComputation<'layout_data> {
                                 self.current_line = ContentSizes::zero();
                                 continue;
                             }
-
-                            let style_text = text_run.parent_style.get_inherited_text();
-                            if style_text.white_space_collapse != WhiteSpaceCollapse::Preserve {
-                                // TODO: need to handle TextWrapMode::Nowrap.
-                                self.line_break_opportunity();
-                                // Discard any leading whitespace in the line. This will always be trimmed.
-                                if self.had_content_yet {
-                                    // Wait to take into account other whitespace until we see more content.
-                                    // Whitespace at the end of the line will always be trimmed.
-                                    self.pending_whitespace += advance;
+                            if !matches!(
+                                style_text.white_space_collapse,
+                                WhiteSpaceCollapse::Preserve | WhiteSpaceCollapse::BreakSpaces
+                            ) {
+                                if can_wrap {
+                                    self.line_break_opportunity();
+                                } else if self.had_content_yet_for_min_content {
+                                    self.pending_whitespace.min_content += advance;
+                                }
+                                if self.had_content_yet_for_max_content {
+                                    self.pending_whitespace.max_content += advance;
                                 }
                                 continue;
                             }
-                            if style_text.text_wrap_mode == TextWrapMode::Wrap {
+                            if can_wrap {
+                                self.pending_whitespace.max_content += advance;
                                 self.commit_pending_whitespace();
                                 self.line_break_opportunity();
-                                self.current_line.max_content += advance;
-                                self.had_content_yet = true;
                                 continue;
                             }
                         }
 
                         self.commit_pending_whitespace();
                         self.add_inline_size(advance);
-                        self.had_content_yet = true;
+                        if can_wrap && run.glyph_store.ends_with_whitespace() {
+                            self.line_break_opportunity();
+                        }
```

Comment:
Does this only happen with `WhiteSpaceCollapse::BreakSpaces`? If so, it might be nice to leave a comment here as that is really what allows the line break opportunity after white space.

Reply:
Yes, the condition can hold in other cases but they are handled above with a `continue`, so by this point the condition should only be true for `WhiteSpaceCollapse::BreakSpaces`.